README.md

```diff
@@ -709,6 +722,149 @@ options on to the custom `TaskQueue` implementation. (Note that because the
 queue options are set as a property on the task, tasks with queue
 options cannot be submitted as JavaScript primitives).
 
+## Custom Worker Scheduling
+
+Piscina provides a default scheduling algorithm that attempts to spawn threads
+based on the max number of threads, and the possibility of a task being cancelled.
+
+This algorithm can be overriden by passing a Custom Scheduler to the `Piscina`
+constructor.
+
+Custom schedulers must implement the `Scheduler` interface, described below:
+```ts
+class TaskScheduler {
+  size: number = 0
+
+  // The constructor will receive the max number of concurrent task per worker
+  // as passed when instantiating Piscina
+  constructor (concurrentTasksPerWorker: number) {
+  }
+
+  [Symbol.iterator] (): IterableIterator<PiscinaWorker> {
+    // Meant to enable Piscina iterate over the workers for different purposes
+  }
+
+  add (worker: PiscinaWorker): void {
+    // Will be called when a new worker is created
+    // Workers will be marked as ready after initialization
+    // You can listen to it calling worker.onReady and scheduling a callback
+  }
+
+  pick (task: PiscinaTask, runOptions: RunTaskOptions): PiscinaWorker | null {
+    // Will be called when a new task is submitted and seeks to be assigned
+    // to a given worker.
+    // Returning null will mean that no worker is available to run the task, and
+    // Piscina will try to create a new one or will put it into the Task Queue
+    // It is up to the scheduler decide how it wants to pick a worker for a given task.
+  }
+
+  onAvailable (cb: (worker: PiscinaWorker) => void): void {
+    // This method is used by Piscina to listen when a new Worker has been marked as ready
+    // after being created.
+  }
+
+  delete (worker: PiscinaWorker): void {
+    // Will be called when a worker must be terminated for different reasons
+  }
+
+  onWorkerAvailable (worker: PiscinaWorker): void {
+    // Will be called whenever a has finished a task and is ready to receive a new one.
```

Comment:
Why no just call add again? and rename pick to shift or something?

Reply:
I think I'd need to better work on the naming, as it might be confusing.

So far they have different purposes; the `add` has just the purpose of keeping track of new spawned threads (basically fresh ones). Meanwhile `onWorkerAvailable` mostly scopes the fact that a Worker has just finished executing a task and might be candidate for a new task; so piscina can continue rolling up more tasks.

wdyt?

Reply:
I think there might be an unnecessary overlap here. Does the scheduler even need to know when a thread is spawned? I think it only needs to know when a thread is available, period. No?

Reply:
i.e. I think you can replace add/delete with workerAvailable/WorkerUnavailable or something.

Reply:
another option is that you skip all of those, and pick receives an array of all available workers.

Reply:
>Does the scheduler even need to know when a thread is spawned? I think it only needs to know when a thread is available, period. No?

Hmm... see your point, but also the value of the Scheduler knowing when a thread is spawned. The idea is to allow the scheduler to have control to manage the threads at all times, so it knows its metrics, usage, and others and can decide which worker/thread to pick next of all available.
Piscina just takes control of spawning and counting that threads are within the boundaries of the options.

i.e. an example of a simple workload
maxThreads:2
minThreads:2

Piscina -> threads <= minThreads -> spawn new thread -> notify scheduler of new thread -> being a new

maxThreads:3
minThreads:2
New incoming task
new task -> threads < maxThreads [yes] -> spawn new worker -> enqueue task
new task -> threads < maxThreads [no] -> Scheduler#pick -> use worker

Task Finished
worker -> piscina [task done] -> scheduler#onWorkerAvailable -> calls cbs to notify piscina a worker is available for new task -> piscina picks task from queue -> loops over new task workflow

But yeah, the naming is bad on the APIs, I'll change them 🤔

Reply:
> so it knows its metrics, usage, and others and can decide which worker/thread to pick next of all available.

What's stopping it from starting to do from when it first becomes avaiable? Just keep a WeakMap<Worker, stats>

Reply:
Yeah, nothing actually 🤔 

But then, do you suggest that is better to just make the Scheduler act as some sort of semaphore that hints Piscina what worker to use next?
And keep all the worker_thread management internally to piscina?

Reply:
That's what I'm think based on this API... the current iteration is a bit in the middle gray zone and is a little ambigious where it's responsibility starts and ends....

Reply:
The problem I guess is whether/how the scheduler tells piscina to create new threads or queue tasks.

Reply:
you could make pick or whatever return `Worker|SPAWN|QUEUE`

Reply:
Yea, this abstraction is tricky

Reply:
Maybe move the queue logic into the scheduler? So that it always either has to accept the task or ask for a new worker?

Reply:
Alternatively we take the `maxThreads` setting?